### ISSUE TITLE
Advanced benchmark measure affected by setup function

### ISSUE THREAD

**Issue by flsilves** [catchorg/Catch2#2584]
**Describe the bug**
When using the `BENCHMARK_ADVANCED` feature in the form of:

```
    BENCHMARK_ADVANCED("Test Function benchmark")
    (Catch::Benchmark::Chronometer meter) {
      setup();
      return meter.measure([&] { sut(); });
    };
```
I realized that depending on the nature of the `setup` function it can affect the output average time measured up to a factor of `2x`. Note that is happens even when `setup` and `sut` are free from interference from each other. Most specifically was observed that this happens more frequently when `setup` involves some form of system call (waiting for a socket or some other system resource, sleep, ...).

**Expected behavior**
Would expect the average time measured by the benchmark to be similar to the one measured by `std::chrono` and to not be affected by the `setup` in the above mentioned conditions. 

**Reproduction steps**
Here is a minimal snippet, a simple `sleep` call of `100ms` was used as `setup`:

```
#include <bits/stdc++.h>
#include <unistd.h>
#include <catch2/benchmark/catch_benchmark.hpp>
#include <catch2/catch_test_macros.hpp>

std::random_device random_device;
std::mt19937 random_engine(random_device());
std::uniform_int_distribution<int> distribution_1_100(1, 100);

constexpr auto N_SAMPLES{1000U};

double test_function() {
  auto rng_vec = std::vector<double>(N_SAMPLES);
  for (auto& e : rng_vec)
    e = distribution_1_100(random_engine);

  return std::accumulate(rng_vec.begin(), rng_vec.end(), 0.0) / static_cast<double>(rng_vec.size());
}

TEST_CASE("Test", "[test]") {
  SECTION("Chrono vs Catch2 Benchmark") {
    std::vector<double> v(N_SAMPLES);
    std::chrono::time_point<std::chrono::high_resolution_clock> start, end;

    BENCHMARK_ADVANCED("Test Function benchmark")
    (Catch::Benchmark::Chronometer meter) {
      usleep(100000);  // sleep for 100 ms
      return meter.measure([&] { test_function(); });
    };

    for (auto& e : v) {
      start = std::chrono::high_resolution_clock::now();
      test_function();
      end = std::chrono::high_resolution_clock::now();
      e = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    }

    double chrono_average =
        std::accumulate(v.begin(), v.end(), 0.0) / static_cast<double>(v.size());
    std::cout << "\n\nstd::chrono elapsed time [ns]: " << chrono_average << std::endl;
  }
}
```
Average time measured with sleep:
 - `Catch2 Benchmark`: 32.21 us
 - `std::chrono` : 17.65 us

Average time measured without sleep: 
 - `Catch2 Benchmark`: 17.61 us
 - `std::chrono` : 17.87 us
 
Notes:
- There's always some variance when running benchmarks from run to run, to minimize this effect some of the advices from google benchmark such as [disabling-cpu-frequency-scaling](https://github.com/google/benchmark/blob/main/docs/user_guide.md#disabling-cpu-frequency-scaling) were done for the above measurements.

- The time measured without the sleep function is what I would expect normally, i.e. a little less than the average measured by `std::chrono` since the benchmark takes into account the overhead of the clock measure. 

- The sleep function is just a replacement for a more complex setup function which waits for some conditions that depend on OS resources. 

Here's also a link to [godbolt](https://godbolt.org/z/hjYsec46s) with the snippet, which was mostly a sanity check to ensure that the test function was not being optimized away.

Compiler flags:
`-O3 -std=c++14 -lCatch2Main`

**Platform information:**

 - OS: **20.04.1-Ubuntu**
 - Compiler+version: **gcc 8.4.0**
 - Catch version: **v3.1.1**

**Comment by Svalorzen** [comment]
Hi, I'm not a Catch2 expert, but I think I have reproduced your error and I might know why this happens.

Note that in your example you have only tested the `chrono` solution without sleep. Putting a CPU to sleep can have consequences, such as possibly changing its power mode, and in general it might be sluggish for a bit after execution restarts.

In your example what happens is that Catch2 executes a `usleep` call before each `test_function()` call, while the `chrono` version never does. A more apt comparison would then be:
```
    for (auto& e : v) {
      usleep(100000);  // sleep for 100 ms
      start = std::chrono::high_resolution_clock::now();
      sum += test_function();
      end = std::chrono::high_resolution_clock::now();
      e = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
    }
```
And if you do, you'll see (at least, on my machine it happens), that now `chrono` lags just as much as Catch.

One additional note: if you are running a statistical analysis, and the number of iterations = 1, it is possible that the estimated run time is actually correct (i.e. does not suffer from interference from the sleep). I'm not sure why this happens exactly, but I thought interesting enough to mention.

**Comment by flsilves** [comment]
@Svalorzen Yes, that was indeed the reason, forgot to come back here and close the issue.
